File: src/main/java/net/minestom/server/command/builder/arguments/relative/ArgumentRelativeVec.java
package net.minestom.server.command.builder.arguments.relative;

import net.minestom.server.command.CommandSender;
import net.minestom.server.command.builder.arguments.Argument;
import net.minestom.server.command.builder.exception.ArgumentSyntaxException;
import net.minestom.server.coordinate.Vec;
import net.minestom.server.utils.StringUtils;
import net.minestom.server.utils.location.RelativeVec;
import org.jetbrains.annotations.NotNull;

import java.util.function.Function;

import static net.minestom.server.utils.location.RelativeVec.CoordinateType.*;

/**
 * Common interface for all the relative location arguments.
 */
abstract class ArgumentRelativeVec extends Argument<RelativeVec> {

    private static final char RELATIVE_CHAR = '~';
    private static final char LOCAL_CHAR = '^';

    public static final int INVALID_NUMBER_COUNT_ERROR = 1;
    public static final int INVALID_NUMBER_ERROR = 2;
    public static final int MIXED_TYPE_ERROR = 3;

    private final int numberCount;

    public ArgumentRelativeVec(@NotNull String id, int numberCount) {
        super(id, true);
        this.numberCount = numberCount;
    }

    abstract Function<String, ? extends Number> getRelativeNumberParser();

    abstract Function<String, ? extends Number> getAbsoluteNumberParser();

    @NotNull
    @Override
    public RelativeVec parse(@NotNull CommandSender sender, @NotNull String input) throws ArgumentSyntaxException {
        final String[] split = input.split(StringUtils.SPACE);
        if (split.length != getNumberCount()) {
            throw new ArgumentSyntaxException("Invalid number of values", input, INVALID_NUMBER_COUNT_ERROR);
        }

        double[] coordinates = new double[split.length];
        boolean[] isRelative = new boolean[split.length];
        boolean isLocalType = false;

        for (int i = 0; i < split.length; i++) {
            final String element = split[i];
            try {
                final char modifierChar = element.charAt(0);

                if (isLocalType && modifierChar != LOCAL_CHAR) {
                    throw new ArgumentSyntaxException("Cannot mix world & local coordinates (everything must either use ^ or not)", input, MIXED_TYPE_ERROR);
                }

                switch (modifierChar) {
                    case LOCAL_CHAR: {
                        isLocalType = true;
                        // Everything in local has to be relative. Fall through.
                    }
                    case RELATIVE_CHAR: {
                        isRelative[i] = true;
                        if (element.length() == 1) break;
                        final String potentialNumber = element.substring(1);
                        coordinates[i] = getRelativeNumberParser().apply(potentialNumber).doubleValue();
                        break;
                    }
                    default: {
                        coordinates[i] = getAbsoluteNumberParser().apply(element).doubleValue();
                        break;
                    }
                }
            } catch (NumberFormatException e) {
                throw new ArgumentSyntaxException("Invalid number", input, INVALID_NUMBER_ERROR);
            }
        }

        final boolean xRelative = isRelative[0];
        final boolean yRelative = split.length == 3 && isRelative[1];
        final boolean zRelative = isRelative[split.length == 3 ? 2 : 1];

        final RelativeVec.CoordinateType type;
        if (isLocalType) {
            type = LOCAL;
        } else if (xRelative || yRelative || zRelative) {
            type = RELATIVE;
        } else {
            type = ABSOLUTE;
        }

        return new RelativeVec(split.length == 3 ?
                new Vec(coordinates[0], coordinates[1], coordinates[2]) : new Vec(coordinates[0], coordinates[1]),
                type,
                xRelative, yRelative, zRelative);
    }

    /**
     * Gets the amount of numbers that this relative location needs.
     *
     * @return the amount of coordinate required
     */
    public int getNumberCount() {
        return numberCount;
    }
}

File: src/main/java/net/minestom/server/command/builder/arguments/relative/ArgumentRelativeVec.kt
package net.minestom.server.command.builder.arguments.relative

import org.jglrxavpok.hephaistos.parser.SNBTParser.parse
import net.minestom.server.command.builder.arguments.number.ArgumentNumber
import java.util.function.BiConsumer
import net.minestom.server.utils.binary.BinaryWriter
import net.minestom.server.command.builder.exception.ArgumentSyntaxException
import java.lang.NumberFormatException
import java.lang.NullPointerException
import net.minestom.server.command.builder.NodeMaker
import net.minestom.server.network.packet.server.play.DeclareCommandsPacket
import java.math.BigDecimal
import net.minestom.server.utils.location.RelativeVec
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeVec
import net.minestom.server.utils.location.RelativeVec.CoordinateType
import net.minestom.server.coordinate.Vec
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentRegistry
import net.minestom.server.entity.EntityType
import net.minestom.server.item.Enchantment
import net.minestom.server.potion.PotionEffect
import java.time.temporal.TemporalUnit
import net.minestom.server.command.builder.arguments.minecraft.ArgumentTime
import it.unimi.dsi.fastutil.chars.CharList
import it.unimi.dsi.fastutil.chars.CharArrayList
import java.lang.IllegalArgumentException
import net.minestom.server.command.builder.arguments.minecraft.ArgumentUUID
import net.kyori.adventure.text.format.NamedTextColor
import net.minestom.server.command.builder.arguments.minecraft.ArgumentColor
import net.minestom.server.command.builder.arguments.minecraft.ArgumentRange
import net.minestom.server.utils.entity.EntityFinder
import net.minestom.server.command.builder.arguments.minecraft.ArgumentEntity
import java.lang.StringBuilder
import net.minestom.server.entity.GameMode
import net.minestom.server.utils.entity.EntityFinder.EntitySort
import net.minestom.server.command.builder.arguments.minecraft.ArgumentIntRange
import java.lang.IllegalStateException
import org.jglrxavpok.hephaistos.nbt.NBT
import org.jglrxavpok.hephaistos.nbt.NBTException
import net.minestom.server.command.builder.arguments.minecraft.ArgumentNbtTag
import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer
import net.minestom.server.command.builder.arguments.minecraft.ArgumentComponent
import net.minestom.server.item.ItemStack
import net.minestom.server.command.builder.arguments.minecraft.ArgumentItemStack
import net.minestom.server.item.Material
import org.jglrxavpok.hephaistos.nbt.NBTCompound
import net.minestom.server.command.builder.arguments.minecraft.ArgumentBlockState
import net.minestom.server.command.builder.arguments.minecraft.ArgumentNbtCompoundTag
import net.minestom.server.command.builder.arguments.minecraft.ArgumentResourceLocation
import net.minestom.server.command.builder.ArgumentCallback
import net.minestom.server.command.builder.suggestion.SuggestionCallback
import net.minestom.server.command.builder.arguments.Argument.ArgumentMap
import net.minestom.server.command.builder.arguments.Argument.ArgumentFilter
import java.util.function.UnaryOperator
import java.util.Locale
import java.lang.SafeVarargs
import net.minestom.server.command.builder.NodeMaker.ConfiguredNodes
import net.minestom.server.command.builder.arguments.number.ArgumentInteger
import net.minestom.server.command.builder.arguments.number.ArgumentDouble
import net.minestom.server.command.builder.arguments.number.ArgumentFloat
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentEnchantment
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentParticle
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentPotionEffect
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentEntityType
import net.minestom.server.command.builder.arguments.minecraft.ArgumentFloatRange
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeBlockPosition
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeVec3
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeVec2
import net.minestom.server.command.builder.arguments.number.ArgumentLong
import net.minestom.server.command.builder.CommandContext
import net.minestom.server.command.builder.parser.ValidSyntaxHolder
import net.minestom.server.command.builder.parser.CommandParser
import net.minestom.server.command.builder.CommandResult
import net.minestom.server.command.builder.CommandDispatcher
import net.minestom.server.MinecraftServer
import net.minestom.server.command.builder.arguments.*
import net.minestom.server.utils.StringUtils
import java.util.Set
import java.util.function.Function

/**
 * Common interface for all the relative location arguments.
 */
abstract class ArgumentRelativeVec(
    id: String,
    /**
     * Gets the amount of numbers that this relative location needs.
     *
     * @return the amount of coordinate required
     */
    val numberCount: Int
) : Argument<RelativeVec>(id, true) {

    abstract val relativeNumberParser: Function<String, out Number>
    abstract val absoluteNumberParser: Function<String, out Number>
    @Throws(ArgumentSyntaxException::class)
    override fun parse(input: String): RelativeVec {
        val split = input.split(StringUtils.SPACE).toTypedArray()
        if (split.size != numberCount) {
            throw ArgumentSyntaxException("Invalid number of values", input, INVALID_NUMBER_COUNT_ERROR)
        }
        val coordinates = DoubleArray(split.size)
        val isRelative = BooleanArray(split.size)
        var type: CoordinateType? = null
        for (i in split.indices) {
            val element = split[i]
            try {
                val modifierChar = element[0]
                if (MODIFIER_CHARS.contains(modifierChar)) {
                    isRelative[i] = true
                    if (type == null) {
                        type = if (modifierChar == LOCAL_CHAR) CoordinateType.LOCAL else CoordinateType.RELATIVE
                    } else if (type != (if (modifierChar == LOCAL_CHAR) CoordinateType.LOCAL else CoordinateType.RELATIVE)) {
                        throw ArgumentSyntaxException(
                            "Cannot mix world & local coordinates (everything must either use ^ or not)",
                            input,
                            MIXED_TYPE_ERROR
                        )
                    }
                    if (element.length > 1) {
                        val potentialNumber = element.substring(1)
                        coordinates[i] = relativeNumberParser.apply(potentialNumber).toDouble()
                    }
                } else {
                    if (type == null) {
                        type = CoordinateType.ABSOLUTE
                    } else if (type == CoordinateType.LOCAL) {
                        throw ArgumentSyntaxException(
                            "Cannot mix world & local coordinates (everything must either use ^ or not)",
                            input,
                            MIXED_TYPE_ERROR
                        )
                    }
                    coordinates[i] = absoluteNumberParser.apply(element).toDouble()
                }
            } catch (e: NumberFormatException) {
                throw ArgumentSyntaxException("Invalid number", input, INVALID_NUMBER_ERROR)
            }
        }
        return RelativeVec(
            if (split.size == 3) Vec(coordinates[0], coordinates[1], coordinates[2]) else Vec(
                coordinates[0],
                coordinates[1]
            ),
            type!!,
            isRelative[0], split.size == 3 && isRelative[1], isRelative[if (split.size == 3) 2 else 1]
        )
    }

    companion object {
        private const val RELATIVE_CHAR = '~'
        private const val LOCAL_CHAR = '^'
        private val MODIFIER_CHARS = Set.of(RELATIVE_CHAR, LOCAL_CHAR)
        const val INVALID_NUMBER_COUNT_ERROR = 1
        const val INVALID_NUMBER_ERROR = 2
        const val MIXED_TYPE_ERROR = 3
    }
}